### PR TITLE
Feature 5094: export settings when activating safe mode

### LINKED ISSUE #5094: Restore settings after Safe Mode
**Is your feature request related to a problem? Please describe.**
Once you enable Safe Mode, there is no way to restore the old settings. This can cause the loss of multiple exclusions across multiple features.

**Describe the solution you'd like**
There is a way to restore the old settings after enabling Safe Mode instead of starting over and create double work in tricky configuration and exclusions.

**Describe alternatives you've considered**
These are the options I  have in mind:
1. Adding a download link to export the current settings in the Safe Mode option in the Deactivation form
image
2. Adding a _Revert Safe Mode_ option in the plugins page:
image
3. Not related to Safe Mode, but the deactivation form could have a **Export settings** link/button, just in case anyone wants to save current settings before deleting WP Rocket.

**Additional context**
Ticket: https://secure.helpscout.net/conversation/1898851114/345290/

Thanks!

### PR DESCRIPTION
**Pull request by CrochetFeve0251**
## Description

Added an option to export options when activating the safe mode.

For that we add an option on the form, a request when in JS when submitting the form and a new REST route.

Fixes #5094
## Type of change

- [ ] New feature (non-breaking change which adds functionality)

## Is the solution different from the one proposed during the grooming?

The solution is different from grooming due to the fact the groomed solution blocked the user into the modal.
We used instead some JS and a rest route to export the options.

## How Has This Been Tested?

- [ ] Test on my local env

# Checklist:

Please delete the options that are not relevant.

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: inc/Engine/Admin/API/Subscriber.php [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+		list( $filename, $options ) = rocket_export_options();
+
+		nocache_headers();
+		@header( 'Content-Type: application/json' ); //phpcs:ignore WordPress.PHP.NoSilencedErrors.Discouraged
+		@header( 'Content-Disposition: attachment; filename="' . $filename . '"' ); //phpcs:ignore WordPress.PHP.NoSilencedErrors.Discouraged
+		@header( 'Content-Transfer-Encoding: binary' ); //phpcs:ignore WordPress.PHP.NoSilencedErrors.Discouraged
+		@header( 'Content-Length: ' . strlen( $options ) ); //phpcs:ignore WordPress.PHP.NoSilencedErrors.Discouraged
+		@header( 'Connection: close' ); //phpcs:ignore WordPress.PHP.NoSilencedErrors.Discouraged
+		echo $options; //phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+		exit();
```
Comment (remyperona) [inline inc/Engine/Admin/API/Subscriber.php:71]: Is this necessary for the JSON output? the REST API will return the formatted response with `rest_ensure_response()`

Comment (CrochetFeve0251) [reply inc/Engine/Admin/API/Subscriber.php:71]: @Tabrisrp can we set a filename to the JSON file with this function?

Comment (remyperona) [reply inc/Engine/Admin/API/Subscriber.php:71]: right, that's true

**Review by remyperona — APPROVED** [review]

**Comment by vmanthos** [comment]
@CrochetFeve0251 I tested this both on rocketlabsqa (clean installation) and on my local Apache installation.

When clicking the "Confirm" button in the modal regardless of whether the checkbox is checked/unchecked:
1. WP Rocket is not deactivated. :white_check_mark: 
2. Safe mode is applied. :white_check_mark: 
3. The settings file is _not_ exported. :x: 

If you select the option to "deactivate and snooze the message" the plugin _is_ deactivated.

Also, part of the wording mentioned in item 6 [here](https://github.com/wp-media/wp-rocket/issues/5094#issuecomment-1311798709) is missing.

Can you please look into this? :pray:

**Comment by vmanthos** [comment]
@CrochetFeve0251 The error I reported previously is fixed. :pray: 

The exported `.json` file when enabling safe mode isn't beautified, contrary to the one that is exported from the tools tab.

Can you please take another look? :pray: 

Also, remember to change the text according to the acceptance criteria:
* item 6 https://github.com/wp-media/wp-rocket/issues/5094#issuecomment-1311798709

**Comment by CrochetFeve0251** [comment]
> @CrochetFeve0251 The error I reported previously is fixed. pray
> 
> The exported `.json` file when enabling safe mode isn't beautified, contrary to the one that is exported from the tools tab.
> 
> Can you please take another look? pray
> 
> Also, remember to change the text according to the acceptance criteria:
> 
>     * item 6 [Restore settings after Safe Mode #5094 (comment)](https://github.com/wp-media/wp-rocket/issues/5094#issuecomment-1311798709)

@vmanthos ok I fixed both issues.

**Comment by vmanthos** [comment]
@CrochetFeve0251 I tested this and the exported file is still minified.

Also, the texting should go next to "and export WP Rocket settings".

File: assets/js/wpr-admin-common.js [hunk 2]
```diff
@@ -4,4 +4,62 @@ jQuery( document ).ready( function( $ ){
...
+
+				const filenames = disposition.match('filename="([^"]+)"');
+
+				if(! filenames.length) {
+					return;
+				}
+
+				const filename = filenames.pop();
+
+				const url = URL.createObjectURL( new Blob( [ JSON.stringify(data, null, 2) ], {
```
Comment (CrochetFeve0251) [inline assets/js/wpr-admin-common.js:45]: @vmanthos can you check if this line is on the website when you test?

Comment (vmanthos) [reply assets/js/wpr-admin-common.js:45]: Indeed, it was an issue with caching.

**Review by vmanthos — APPROVED** [review]
Working as expected. :+1:

**Pull request merged**